lib/SILOptimizer/Transforms/PerformanceInliner.cpp

```diff
@@ -1125,8 +1126,25 @@ SILFunction *SILPerformanceInliner::getEligibleFunction(FullApplySite AI) {
 
   // We don't support this yet.
   if (AI.hasSubstitutions()) {
-    if (!EnableSILInliningOfGenerics)
+    // Do not inline @_semantics functions when compiling the stdlib,
+    // because they need to be preserved, so that the optimizer
+    // can properly optimize a user code later.
+    auto ModuleName = Callee->getModule().getSwiftModule()->getName().str();
+    if (Callee->hasSemanticsAttrThatStartsWith("array.") &&
+        (ModuleName == STDLIB_NAME || ModuleName == SWIFT_ONONE_SUPPORT))
       return nullptr;
+
+    // Always inline generic functions which are marked as
+    // AlwaysInline or transparent.
+    if (Callee->getInlineStrategy() == AlwaysInline ||
+        Callee->isTransparent()) {
+      DEBUG(llvm::dbgs() << "Inline a generic function " << Callee->getName()
+                         << "\n");
+    } else {
+      // Otherwise, only inline generic functions if asked to do so.
+      if (!EnableSILInliningOfGenerics)
```

Comment:
I'm not sure, because there is no early exit here. More over, when I tried it, the code looked uglier and less understandable.